components/Forms/SignUp.js

```diff
@@ -0,0 +1,187 @@
+import { useReducer } from 'react';
+import Link from 'next/link';
+
+import Button from '@mui/material/Button';
+import CssBaseline from '@mui/material/CssBaseline';
+import TextField from '@mui/material/TextField';
+import Grid from '@mui/material/Grid';
+import Box from '@mui/material/Box';
+import { Link as MUILink } from '@mui/material';
+import Typography from '@mui/material/Typography';
+import Container from '@mui/material/Container';
+
+const formReducer = (state, action) => {
+	switch (action.type) {
+		case 'FORM_CHANGE':
+			const newState = {
+				controls: {
+					...state.controls,
+					[action.control]: {
+						value: action.value,
+						isValid: action.isValid,
+						isTouched: true,
+					},
+				},
+			};
+			newState.isValid = Object.values(newState.controls).every(
+				control => control.isValid,
+			);
+			return newState;
+		default:
+			return state;
+	}
+};
+
+const SignUp = () => {
+	const [form, dispatch] = useReducer(formReducer, {
```

Comment:
A lot of code related to the state management and validation is repetitive such as the reducer and updating of touched state and deciding when to show an error. Can we abstract this into a useForm which can be used for any form? This hook would take in the initial value of each control, a validator function and provide the current value, error, isTouched and such details for each form control.
Alternative: Use a prebuilt solution such as https://react-hook-form.com/. If you choose this route, this is probably how you would need to integrate it with MUI components https://react-hook-form.com/get-started#IntegratingwithUIlibraries.

Reply:
Agreed, I myself didn't like how I did validation I will refactor all of this into a custom useForm hook.